### PR TITLE
Bs-11 User and Alarm Progress

### PR DESCRIPTION
**Pull request by andrchoi**
Added functions to add user and alarm data in metrics database.

### CODE REVIEW

**Comment by andrchoi** [comment]
Should the MetricServices be in the Services folder? 

Also is there a way you wanted to create the models? Knex is only a query builder and I couldn't find any documentation for that. I found a package [sequelize](https://sequelize.org/v5/manual/models-definition.html) that can create models, however if it matters, sequelize overlaps with knex because they both query.

**Comment by Asiapenguin** [comment]
1. yes
2. Ahh sequelize. Stick with knex. The model could just be a javascript object, no restrictions.

File: README.md [hunk 1]
```diff
@@ -138,3 +158,99 @@ HTTP Response:
...
+
+PUT "/metrics/alarm"
+
+Request Body:
+
+```
+{
+    "alarmInfo": [
+        {
+            "alarmID": integer,
```
Comment (Asiapenguin) [inline README.md]: alarmId should be a path param. It identifies a resource, not sort/filter

File: README.md [hunk 2]
```diff
@@ -138,3 +158,99 @@ HTTP Response:
...
+```
+
+HTTP Response:
+
+```
+{
+    "response": [
+        {
+            "success": boolean,
+            "message": string
```
Comment (Asiapenguin) [inline README.md]: Return the updated sentStatus, no need for message

File: README.md (hunk 1 above)
Comment (Asiapenguin) [reply README.md]: Is there a reason why you put alarmInfo wrapping the actual body data that we are interested in?

File: README.md (hunk 2 above)
Comment (Asiapenguin) [reply README.md]: Same here with the "response", any reason?

File: README.md [hunk 3]
```diff
@@ -138,3 +158,99 @@ HTTP Response:
...
+        }
+        ...
+    ]
+}
+```
+
+## Update Alarm Log End Time
+
+PUT "/metrics/alarm/extend"
```
Comment (Asiapenguin) [inline README.md]: This should be the same PUT endpoint to "Update Alarm Log Sent Status". The way you can detect which is which is by looking at the request body. If the attribute exists there, update it.

File: server.js [hunk 4]
```diff
@@ -32,6 +33,10 @@ app.get("/users/:id/responders", middleware.checkToken, user.getResponders);
 app.get("/users/:id/responders/count",middleware.checkToken,user.getResponderCount);
 app.delete("/users/:id/responders/:responderid", middleware.checkToken, user.deleteResponder);
 
+app.post("/metrics/alarm", alarmService.alarmStart)
```
Comment (Asiapenguin) [inline server.js]: Do not use the service methods directly. Go through an Alarm Controller. The controller will be able to do all the validation checking on the request bodies and catch any errors thrown by your service methods. Let me know if this is not clear.

File: Controllers/alarm.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    if (userID.length === 1) {
+      userID = userID[0].id;
+    } else {
+      handle.notFound(res, "Cannot find userID for username")
+    }
+
+    let alarmID = await alarmService.createAlarmLog(userID, data.startTime, data.endTime);
+    console.log(alarmID);
+    res.status(200).json({
+      alarmID: alarmID,
```
Comment (Asiapenguin) [inline Controllers/alarm.js]: change alarmID to id

File: Models/metrics/alarm.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+const alarm = {
```
Comment (Asiapenguin) [inline Models/metrics/alarm.js]: Look into creating a Javascript class, such that you can instantiate an object by using "new" with parameters. Create these objects in your service before you do insert/update/delete and refer to their attributes

File: README.md [hunk 7]
```diff
@@ -138,3 +158,71 @@ HTTP Response:
...
+
+## Adding Alarm Metric Log
+
+POST "/metrics/alarm"
+
+Request Body:
+
+```
+{
+    "alarmInfo": [
```
Comment (Asiapenguin) [inline README.md]: There is no batch POST, so you dont need an array, just have username startTime and endTime as root attributes to the resquest body

File: README.md [hunk 8]
```diff
@@ -138,3 +158,71 @@ HTTP Response:
...
+        ...
+    ]
+}
+```
+
+HTTP Response:
+
+```
+{
+    "addedAlarm": [
```
Comment (Asiapenguin) [inline README.md]: Same with response body.

File: Services/alarmMetricService.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
+let model = require("../Models/metrics/alarm").alarm;
+
+let metricDB = require("knex")({
```
Comment (Asiapenguin) [inline Services/alarmMetricService.js]: Extract into a singleton, look at database.js and server.js. It connects on startup and all other services that use the postgres database will get the instance from the singleton

File: Services/alarmMetricService.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+      alarmsent: "FALSE"
+    }).returning("*").then(res => {
+      console.log(res);
+      alarmLogID = res[0].id;
+    });
+
+    return alarmLogID;
+
+  } catch (err) {
+    throw err;
```
Comment (Asiapenguin) [inline Services/alarmMetricService.js]: Add the error message here instead

File: Services/userMetricService.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    await metricDB("users").where({
+      username: username
+    }).update({
+      lastlogin: metricDB.fn.now()
+    }).returning("*").then(res => {
+      checkExists = res;
+      console.log(checkExists);
+    });
+
+    if (checkExists.length < 1) {
```
Comment (Asiapenguin) [inline Services/userMetricService.js]: checkExists might not have updated when this line is executed because it is not within the .then block

**Comment by Asiapenguin** [comment]
@andrchoi Let me know when you have understood the code and I will merge

**Comment by andrchoi** [comment]
I looked through it, the changes make sense. I'll see if I can get users fixed in the next pr

**Pull request merged**